File: problem_7/Solution.scala
object Solution {
  def main(args: Array[String]) {
    tests()
    println(nPrime(10001))
  }

  def tests() = {
    assert(isPrime(1) == false)
    assert(isPrime(2) == true)
    assert(isPrime(4) == false)
    assert(isPrime(5) == true)
    assert(isPrime(8) == false)
    assert(isPrime(22) == false)
    assert(isPrime(29) == true)
  }

  def nPrime(number:Int): Int = {
    var primeCounter = 1
    var i = 1
    var nprime = 0
    do {
      if(isPrime(i)) {
        primeCounter += 1
        nprime = i
      }
      i += 2
    } while (primeCounter != number)
    nprime
  }

  def isPrime(n:Int): Boolean = {
    if(n==1) return false
    if(n==2) return true
    if(n%2 ==0) return false

    val sqRt = scala.math.sqrt(n).toInt
    var i = 3
    while(i<=sqRt) {
      if(n%i == 0) return false
      i += 2
    }
    return true
  }
}

File: problem_7/Solution.java
/*
   By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13, we can see that the 6th prime is 13.
   What is the 10 001st prime number?
*/

class Solution {
  public static void main(String[] args) {
    assert(isPrime(2) == true);
    assert(isPrime(3) == true);
    assert(isPrime(4) == false);
    assert(isPrime(5) == true);

    System.out.println(nPrime(10001));
  }

  public static int nPrime(int n) {
    int primesCounter = 1;
    int i = 1;
    int largest_prime = 0;
    while(primesCounter != n) {
      if(isPrime(i)) {
        primesCounter++;
        largest_prime = i;
      }
      i+=2;
    }
    return largest_prime;
  }

  public static boolean isPrime(int number) {
    if (number ==1) return false;
    if (number ==2) return true;
    if (number % 2 == 0) return false;
    int sq_rt = (int) Math.sqrt(number);
    int i = 3;
    while(i<=sq_rt) {
      if (number % i == 0) return false;
      i += 2;
    }
    return true;
  }
}
